Teach miopen-dialect to lower bf16 end-to-end. Take 2.

whchung opened the pull request:
2nd take to merge bf16 logic to `miopen-dialect`.

Internally, `i16` is used to represent `bf16` so XDLOPS intrinsics can be directly applied.

whchung commented:
@asleepzzz some unit test failures in CI:

http://ml-ci.amd.com:21096/job/MLIR/job/MLIR-downstream-unit-test/2886/console

4 of them:

```
  MLIR :: CAPI/ir.c
  MLIR :: Conversion/GPUToROCDL/gpu-to-rocdl.mlir
  MLIR :: Conversion/StandardToSPIRV/std-types-to-spirv.mlir
  MLIR :: Dialect/SPIRV/IR/types.mlir
  ```

whchung commented:
@asleepzzz Please fix `git clang-format` checks.

whchung commented:
@asleepzzz , #66 would likely require changes in this PR too. Try make every effort to get this PR merged ASAP, so we can look into other tasks to make `miopen-dialect` more robust.

whchung committed:
Squashed commit after git clang-format.

Revert "Refactor conv solvers (#60)" (#70) (commit: 8b29e75) (details /
githubweb)
add gpu cast fp32 to bf16 op and lowering to llvm (commit: daa986a)
(details / githubweb)
affixtunning lowering failed ,need to fix (commit: dc89814) (details /
githubweb)
lowering to llvm ok (commit: 63e100d) (details / githubweb)
change i16 part ,lowering still failed (commit: ebb87ae) (details /
githubweb)
fix lowering fail (commit: 89672f1) (details / githubweb)
add bf16 convert (commit: 2bb9816) (details / githubweb)
add wrapper bf16 convert (commit: 02f8cc7) (details / githubweb)
add convert (commit: 87f7cd0) (details / githubweb)
add matrixA float32 cpu verification buffer when data is bf16 (commit:
25b0f3e) (details / githubweb)
fix mcpuMemBF16ConvertFloat and add AB matrix fp32 alloc when bf16
(commit: 6331c43) (details / githubweb)
fix bf16 to float, verification not complete (commit: e952477) (details
/ githubweb)
add verification, bf16 and others should be different at next commit
(commit: d55ae3e) (details / githubweb)
add verification (commit: 965bf54) (details / githubweb)
fix ci test: std to spirv bf16 (commit: 9ed212e) (details / githubweb)
remove imms when bf16 now, will fix it (commit: c1b72fa) (details /
githubweb)
fix gpu-to-rocdl.mlir (commit: 8c20166) (details / githubweb)
Fix 3 unit tests. (commit: a44d6c7) (details / githubweb)
Fix mlir-miopen-driver for i16(bf16) type conversion and its test.
(commit: 3afc442) (details / githubweb)
Fix build failures. (commit: e3e4245) (details / githubweb)

whchung commented:
@jerryyin ping. All unit tests are passing. `git clang-format` shows `LowerMIOpenOps.h` needs to be fixed. However on my local box `git clang-format` returns no change. Could you help expedite merging the PR so @asleepzzz could be freed from the task? Thanks.

jerryyin commented:
@whchung I believe the default clang-format is not the one that CI adopted. CI is using clang-format from `/opt/rocm/llvm/bin/clang-format`, just FYI

jerryyin commented:
WIP in clang-tidy errors

jerryyin commented:
Abstain from code review as I'm not on top of the changes of this PR. @sjw36 Would you like to take a look at this PR?

I can approve if this turns out to block Jack's additional changes.

whchung commented on mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:
```diff
@@ -1213,64 +1257,266 @@ static LogicalResult populateValidationLogic(
   mlir::FuncOp verifyFuncOp;
   if (operation.getValue() == "conv2d") {
-    cpuResults = cpuOutputHostAllocOp;
+    // cpuResults = cpuOutputHostAllocOp;
```
@asleepzzz remove this unused code.

asleepzzz replied:
already remove it

whchung commented on mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:
```diff
@@ -1213,64 +1257,266 @@ static LogicalResult populateValidationLogic(
                                       cpuResults, gpuResults);
   } else if (operation.getValue() == "conv2d_bwd_data") {
-    cpuResults = cpuInputHostAllocOp;
+    // cpuResults = cpuInputHostAllocOp;
```
@asleepzzz remove this unused code.

asleepzzz replied:
already remove it

whchung commented on mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:
```diff
@@ -1213,64 +1257,266 @@ static LogicalResult populateValidationLogic(
                                       cpuResults, gpuResults);
   } else if (operation.getValue() == "conv2d_bwd_weight") {
-    cpuResults = cpuFilterHostAllocOp;
+    // cpuResults = cpuFilterHostAllocOp;
```
@asleepzzz remove this unused code.

asleepzzz replied:
already remove it

sjw36 commented on mlir/include/mlir/Dialect/GPU/GPUOps.td:
```diff
@@ -925,4 +925,16 @@ def GPU_MFMAOp:
+    Results<(outs I16:$out)> {
+  let summary = "from fp32 to bf16";
+  let description = [{
+    The `gpu.bf_convert` op change fp32 to bf16, but we store with i16.
```
Comment cleanup:

> The 'gpu.bf_convert' op converts an fp32 value to bf16 value. Stored as i16 for usage compatibility.

asleepzzz replied:
> @whchung I believe the default clang-format is not the one that CI adopted. CI is using clang-format from `/opt/rocm/llvm/bin/clang-format`, just FYI

asleepzzz replied:
already changed comments

whchung commented:
@sjw36 ping.

sjw36 reviewed (commented):
> @sjw36 ping.

In a meeting.. will finish review after 2.

sjw36 commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -4610,12 +4617,23 @@ struct XdlopsGemmV2RewritePattern
       VectorType argVectorType = argType.template dyn_cast<VectorType>();
-      KRepeats = (dataType.getWidth() / 8 * argVectorType.getShape()[0]) / (dataType.getWidth() / 8 * k_base);
+      KRepeats =
+          (dataType.template dyn_cast<IntegerType>().getWidth() / 8 *
```
Factor out dataType byte-width. And it should cancel out anyway right?

asleepzzz replied:
yes, already remove it

sjw36 commented on mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp:
```diff
@@ -180,42 +180,79 @@ struct MFMAOpLowering : ConvertToLLVMPattern {
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto bfOp = cast<gpu::BFConvertOp>(op);
```
Would be good to have a comment with the math.

asleepzzz replied:
added

sjw36 commented on mlir/lib/Conversion/MIOpenToGPU/MIOpenToGPU.cpp:
```diff
@@ -214,6 +214,16 @@ void LowerMIOpenOpsToGPUPass::runOnOperation() {
+        auto loc = op.getLoc();
+        OpBuilder b(op.getContext());
+        b.setInsertionPoint(op);
+        Value cast =
```
Is DataConvert always to BF16?

asleepzzz replied:
yes, DataConvertOp always to BF16

sjw36 commented on mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:
```diff
@@ -31,6 +31,18 @@ int32_t reportErrorIfAny(hipError_t result, const char *where) {
+
+static float bfloat16_to_float(ushort src_val) {
+  bf16_fp32_cvt_t target_val;
+  target_val.ushortvec[1] = src_val;
```
need to assign lower bits (ushortvec[0] = 0).

asleepzzz replied:
assigned

asleepzzz replied:
added it

whchung commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -296,19 +296,28 @@ def MIOpen_MFMAV2Op:
+    MIOpen_Op<"data_convert">,
+    Arguments<(ins F32:$in)>,
+    Results<(outs I16:$out)> {
+  let summary = "from data type to another";
```
Documentation for this op needs to be improved.

asleepzzz replied:
already changed

whchung commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -296,19 +296,28 @@ def MIOpen_MFMAV2Op:
+    Results<(outs I16:$out)> {
+  let summary = "from data type to another";
+  let description = [{
+    The `miopen.data_convert` op transforms some data type.
```
Please state more clearly what this op does actually.

asleepzzz replied:
already changed

sjw36 reviewed (approved):
Looks good. Just a few minor comments.

whchung commented:
@asleepzzz Some minor changes are needed and the PR is good to be taken. Let's try aim for getting this done by Tuesday Taiwan time.

whchung committed:
Fix miopen.fill / miopen.threadwise_gemm in fp16 / bf16 cases. (#76)

* Fix how miopen.fill is used in fp16 / bf16 cases.

Add sanity tests to ensure each lowering step produces valid MLIR.

* Fix how miopen.threadwise_gemm gets lowered for fp16 / bf16 data types.

Augment sanity tests.

* Disable several test configs.

These configs pass on my local box but fails weirdly on CI. Disable them
for now.

* Fix clang-format.

* Break sanity.mlir into 2 files.

One for non-XDLOPS, one for XDLOPS path.

asleepzzz committed:
fix PR comments

asleepzzz committed:
add v1 but not implement yet

whchung commented:
@asleepzzz Please fix 2 failing sanity tests:

```
Failed Tests (2):
  MLIR :: mlir-miopen-driver/sanity.mlir
  MLIR :: mlir-miopen-driver/sanity_pv.mlir
```

whchung commented:
`sanity_pv.mlir` shows:

```
/home/jenkins/workspace/MLIR/MLIR-downstream-unit-test/build/tools/mlir/test/mlir-miopen-driver/Output/sanity_pv.mlir.script: line 3: 22491 Segmentation fault      (core dumped) mlir-miopen-driver -p -pv -t bf16 -c
```

For now please disable its RUN command.


`sanity.mlir` shows:

```
error: 'miopen.threadwise_gemm' op operand #2 must be 2D memref of 32-bit float or 16-bit float values, but got 'memref<8x8xi16, 5>'
Lowering failed.
```

This should be an easy fix. And it suggests when you guys implement bf16 logic you guys didn't check non-XDLOPS path.

whchung commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -269,9 +269,9 @@ def MIOpen_BlockwiseGemmV2Op:
-                   MemRefRankOf<[F32, F16, BF16], [2]>:$matrixC)> {
+    Arguments<(ins MemRefRankOf<[F32, F16, I16], [2]>:$matrixA,
+                   MemRefRankOf<[F32, F16, I16], [2]>:$matrixB,
+                   MemRefRankOf<[F32, F16], [2]>:$matrixC)> {
```
@asleepzzz This line is missing `I16`.

asleepzzz committed:
add output i16 for threadwise

whchung commented on mlir/tools/mlir-miopen-driver/bf16convert.hpp:
```diff
@@ -0,0 +1,41 @@
+
+float bfloat16_to_float(uint16_t src_val) {
+  cvt_bf16_fp32_t target_val;
+  target_val.ushortvec[1] = src_val;
```
```
target_val.ushortvec[0] = 0;
```
?

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -3620,15 +3624,19 @@ struct ThreadwiseGemmRewritePattern
 
+    Value mul = b.create<MulFOp>(loc, b.getF32Type(), gemmAKM, gemmBKN);
+    if (dataType == b.getIntegerType(16))
+      mul = b.create<MulIOp>(loc, b.getIntegerType(16), gemmAKM, gemmBKN);
```
this logic is dangerous. you are adding dangling values into the compiled code. also, it doesn't handle f16 case.

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -3620,15 +3624,19 @@ struct ThreadwiseGemmRewritePattern
-    auto store = b.create<AffineStoreOp>(loc, add, gemmC, memIndicesMN);
+    Value add = b.create<AddFOp>(loc, b.getF32Type(), mul, gemmCMN);
+    if (dataType == b.getIntegerType(16))
+      add = b.create<AddIOp>(loc, b.getIntegerType(16), mul, gemmCMN);
```
same here.

asleepzzz committed:
fix non xdlop can't lowering bf16

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -3620,13 +3636,17 @@ struct ThreadwiseGemmRewritePattern
     auto gemmBKN = b.create<AffineLoadOp>(loc, gemmB, memIndicesKN);
-    auto mul = b.create<MulFOp>(loc, b.getF32Type(), gemmAKM, gemmBKN);
 
+    Value mul = b.create<MulFOp>(loc, dataType, gemmAKM, gemmBKN);
```
incorrect code in case data type is an integer.

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -3620,13 +3636,17 @@ struct ThreadwiseGemmRewritePattern
     auto gemmCMN = b.create<AffineLoadOp>(loc, gemmC, memIndicesMN);
 
-    auto add = b.create<AddFOp>(loc, b.getF32Type(), mul, gemmCMN);
+    Value add = b.create<AddFOp>(loc, dataType, mul, gemmCMN);
```
incorrect code in case data type is an integer.

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -1369,13 +1369,26 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
+      zeroConstantFloatOp =
+          b.create<ConstantFloatOp>(loc, APFloat(0.0f), b.getF32Type());
+    } else if (elementType == b.getF16Type() ||
+               elementType == b.getBF16Type()) {
```
do we still need `getBF16Type()` check here?

asleepzzz committed:
fix non xdlops step3 failed

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-new-bf16

asleepzzz committed:
remove additional op

whchung committed:
Remove unused bf16 type check.
